### ISSUE TITLE
Internal error for vectorize=on and multiplication of INT4 values

### ISSUE THREAD

**Issue by mrigger** [cockroachdb/cockroach#46714]
Consider the following statements:

```sql
SET vectorize=on;
CREATE TABLE t0(c0 INT4);
CREATE TABLE t1(c0 INT4);
INSERT INTO t0(c0) VALUES(0);
INSERT INTO t1(c0) VALUES (0);
SELECT 1 FROM t0, t1 GROUP BY t0.c0 * t1.c0; -- unexpected error from the vectorized runtime: interface conversion: coldata.column is []int32, not []int64
```
Unexpectedly, the `SELECT` results in an internal error with the following stack trace:
```
ERROR: internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []int32, not []int64
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:92: func1()
runtime/panic.go:679: gopanic()
runtime/iface.go:255: panicdottypeE()
github.com/cockroachdb/cockroach/pkg/col/coldata/vec.go:200: Int64()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hash_utils.eg.go:453: rehash()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashtable.go:397: computeBuckets()
github.com/cockroachdb/cockroach/pkg/sql/colexec/hashtable.go:246: build()
github.com/cockroachdb/cockroach/pkg/sql/colexec/unordered_distinct.go:80: Next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/const.eg.go:322: Next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:120: Next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149: next()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140: nextAdapter()
github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:95: CatchVectorizedRuntimeError()
github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177: Next()
github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170: Run()
github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:749: Run()
github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:370: Run()
github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:405: Run()
github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1019: PlanAndRun()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:877: execWithDistSQLEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:770: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:475: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1368: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1297: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:476: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:594: func1()
runtime/asm_amd64.s:1357: goexit()
```
This bug might be related to the open bug https://github.com/cockroachdb/cockroach/issues/46503. However, I assume they are different, because the bug reported in this issue does not require `DISTSQL` to be off.

Environment:
```
Build Tag:    v20.1.0-beta.2-1175-g92495ffb21
Build Time:   2020/03/29 09:01:03
Distribution: CCL
Platform:     linux amd64 (x86_64-linux-gnu)
Go Version:   go1.13.4
C Compiler:   gcc 8.3.0
Build SHA-1:  92495ffb2182671b3338a8c6cd2734cb6de5d392
Build Type:   development
```

**Comment by asubiotto** [comment]
Possible that this could be fixed by #46712

**Comment by yuzefovich** [comment]
This problem was initially unrelated to what #46712 is trying to fix, but I hit a similar issue while working on that PR, so I included in there a second commit to fix this problem as well.
